### DISCUSSION TITLE
static func 에 대한 고찰에서 파생된 static var v.s static let

### DISCUSSION THREAD

**Discussion by seongmin221** [GEON-PPANG/GEON-PPANG-iOS#267] [Ideas]
## 서론

현재 저희 디자인 시스템은 아래와 같이 `static let`으로 구현되어 있습니다 !

```swift
extension UIImage {

    // ...

    // Bread Types
    static let disabledCakeIcon = UIImage(named: "ic_cake_disable_20px")!.withRenderingMode(.alwaysOriginal)
    static let disabledHardIcon = UIImage(named: "ic_hard_disable_20px")!.withRenderingMode(.alwaysOriginal)
    static let disabledSandwichIcon = UIImage(named: "ic_sandwich_disable_20px")!.withRenderingMode(.alwaysOriginal)
    static let enabledCakeIcon = UIImage(named: "ic_cake_enable_20px")!.withRenderingMode(.alwaysOriginal)
    static let enabledHardIcon = UIImage(named: "ic_hard_enable_20px")!.withRenderingMode(.alwaysOriginal)
    static let enabledSandwichIcon = UIImage(named: "ic_sandwich_enable_20px")!.withRenderingMode(.alwaysOriginal)

    // ...
}
```

하지만 당시에 디자인 시스템 이것저것을 참고하면서 어떤게 좋을까 고민을 좀 했었는데요,
그 때는 좀 깊게 공부하지 않아서 기존에 사용하던 `static let` 으로 구현을 했습니다

하지만 이번에 #266 에서 구글링하는 과정을 통해 `static let` 과 `static var` 의 차이점에 대해 더 깊게 알게 되었고,
디자인 시스템에 `static var`과 `computed property` 를 사용하는게 더 좋지 않을까 싶어 아이디어 남깁니다 !

<br>

<br>

## `static let` v.s `static var`

- static 하게 선언된 요소들은 모두 lazy 가 내장(?)되어 있음

### `static let`
- 따라서 한 번 사용하기 전에는 메모리에 올라가지 않음
- 하지만 한 번 사용하면 사용하지 않을 때에도 메모리에 올라가 있음

### `static var` & `computed property`
- `computed property`이기 때문에, `입력값이 없는 함수처럼 동작`
- 함수이므로, 메모리에 저장되지 않음
- 하지만 사용될 때마다 `계산`되어야 함
- 디자인 시스템의 경우 리소스를 사용할 때마다 번들에 접근해야 함 ( CPU의 사용일듯 ? )

<br>

<br>

## 의견

> `static var` & `computed property` 로 사용하는거 어떨지 ~?
> 사실 큰 성능 차이는 없을 것 같긴 하지만...
> 우리가 메모리 사용량이 꽤나 많은 편인 것 같기 때문에 이런 방법도 좋지 않을까라는 생각 ...

실제 변화가 얼마나 있는지는 함 해보고 테스트 해봐야 할 듯 !

---

출처: 
- https://stackoverflow.com/questions/69363771/what-is-the-lifespan-of-a-static-variable-in-swift
- https://ios-development.tistory.com/686#comment21809064

**Comment by jeongdung-eo** [comment]
👍 좋아욤! 성능과 메모리 관점에서 살펴보았을 때, 우리 프로젝트에서는` 메모리 최적화`를 우선시해도 좋을 것 같슴당
